### ISSUE TITLE
Access MS-SQL instance with windows authentication using 'vertx-mssql-client' in a quarkus reactive project

### ISSUE THREAD

**Issue by git4rputuval** [eclipse-vertx/vertx-sql-client#1407]
Our quarkus version is 2.7.2.Final and mssql version is 4.2.4 (using the vertx-mssql-client-4.2.4 jar)  

NTLM authentication applies for jdbc drivers only, and our entire application is built on top of Quarkus Reactive. Going back to JDBC based implementation is another 6 months project, and we have already completed UAT. Going back is not possible. Is there anyway we can get this working?

All our clients are undergoing security audits and pressing on windows authentication feature for App Server to DB connection. Sooner or later, all the projects implementing Quarkus Reactive is going to take a hit, if implementing in banking projects.

So we would request you to provide for a configuration supporting windows authentication or guide us with any work around which could provide with the same result.

**Comment by git4rputuval** [comment]
Can you let us know a date that we can communicate to our clients?

**Comment by tsegismont** [comment]
I'm afraid there is no planned date, even though we'd like to have this feature. We have neither the knowledge of the protocol nor the resources required to test its implementation.

If you're willing to contribute it, I would be glad to help.

Otherwise, I can suggest some alternatives: switching to `vertx-jdbc-client`, or given you're a Hibernate user, switching to a blocking persistence layer.

**Comment by git4rputuval** [comment]
> I'm afraid there is no planned date, even though we'd like to have this feature. We have neither the knowledge of the protocol nor the resources required to test its implementation.
> 
> If you're willing to contribute it, I would be glad to help.
> 
> Otherwise, I can suggest some alternatives: switching to `vertx-jdbc-client`, or given you're a Hibernate user, switching to a blocking persistence layer.

Hi,

We are creating a project that contains a sample representative of our application.  Once that is done, it will be shared shortly. Please let us know if we are making any mistake in creating a blocking persistence layer.

In the meantime, do let us know how to contribute. We will discuss with my development team and work out our timelines to see if we are able to understand, contribute and deliver to our clients at the earliest.

Kind regards,

**Comment by petarov** [comment]
That's also the reason I'm currently not using the vertx-mssql-client currently: 99% of my customers use Windows Server and they mostly prefer named instances with Windows authentication as opposed to classic user/pwd SQL Server authentication. 

I'm using the mssql-jdbc driver with blocking calls and it's doing quite alright so far. One idea that I have, is to possibly switch to a dedicated worker verticle with a `ThreadingModel.VIRTUAL_THREAD` threading model. I wonder if that'd yield any benefits.

In my case, I'm using the `mssql-jdbc_auth-<version>-<arch>.dll` that provides the required Kerberos auth routines, however and as described my Microsoft [1], it should be possible to use the **Krb5LoginModule** [2] as well, and I believe this could be in fact the best possible variant to add Windows authentication support to the vertx-mssql-client.

I hope I can get some time to look for a way to help adding this in the future. It'd be an interesting project.

1 - https://learn.microsoft.com/en-us/sql/connect/jdbc/using-kerberos-integrated-authentication-to-connect-to-sql-server
2 - https://docs.oracle.com/javase/8/docs/jre/api/security/jaas/spec/com/sun/security/auth/module/Krb5LoginModule.html

**Comment by git4rputuval** [comment]
> I'm using the mssql-jdbc driver with blocking calls and it's doing quite alright so far. 

Can you provide the code and/steps to implement the Kerberos auth routines?

**Comment by git4rputuval** [comment]
We are attaching a demo project which demonstrates our application architecture to a small extent to reproduce the errors we're getting. This project attachment is giving a runtime error whereas in our actual application its a compilation error, we are attaching both errors as snapshots, we expect that if either of the error is fixed the other would also be solved.

Below is the demo application
[persistenceUnit_Error.zip](https://github.com/eclipse-vertx/vertx-sql-client/files/14402705/persistenceUnit_Error.zip)

the compilation error in our actual application is 
<img width="960" alt="compilationErrorForSimulatenousDataSources" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/6c3c4348-b4f0-4a46-b989-a9c11d1381fe">

the runtime error wrt this demo application is
<img width="960" alt="runtimeErrorWithSimlutaneousDataSources" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/3473fca7-d385-4609-8124-20bc13535b42">

this is runtime error when we disable reactive datasource using quarkus.datasource.reactive=false in our application.properties file
<img width="725" alt="disableReactiveDataSource" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/cb72f0e4-8602-4708-99ac-0fdbf1e9d355">

<img width="960" alt="runtimeErrorWith ReactiveDisabled" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/fce78253-f4b3-423b-9910-06d691b9a4c2">

The solutions we used by far involved using @PersistenceUnit to attach our entity/model to a specific data source, made two different entities one to map if persistence is done using jdbc and depending on the flag value set in application.properties persistence through our data service layer would either happen from jdbc persistence side (jdbcEntity,jdbcService,jdbcRepo classes) else the usual/preffered reactive side.

**Comment by git4rputuval** [comment]
> I'm afraid there is no planned date, even though we'd like to have this feature. We have neither the knowledge of the protocol nor the resources required to test its implementation.
> 
> If you're willing to contribute it, I would be glad to help.
> 
> Otherwise, I can suggest some alternatives: switching to `vertx-jdbc-client`, or given you're a Hibernate user, switching to a blocking persistence layer.

Hi @tsegismont ,

Please provide the blocking persistence layer option. Let me evaluate this option.

Thanks & Regards,

**Comment by git4rputuval** [comment]
@tsegismont 

Please provide the details on how to contribute. We would like to evaluate it with our existing timelines.

Kind Regards,

**Comment by tsegismont** [comment]
@git4rputuval the firsts steps would be to:

-  clarify which auth scheme(s) you plan to contribute (I'm a bit lost between NTLM, Kerberos ...)
- verify how the contribution can be tested (we need to be able to test it to support it)

**Comment by mswatosh** [comment]
>verify how the contribution can be tested

As someone who has dealt with Kerberos and SQLServer, last I tried, SQLServer requires an Active Directory environment when using Kerberos. This makes it difficult to test with TestContainers, since even if the SQLServer instance is on Linux, the AD domain controller needs to run on windows. It may be possible to get it working with Samba's AD domain controller, but I couldn't justify the time investment to get it working.

**Comment by git4rputuval** [comment]
> @git4rputuval the firsts steps would be to:
> 
> * clarify which auth scheme(s) you plan to contribute (I'm a bit lost between NTLM, Kerberos ...)
> * verify how the contribution can be tested (we need to be able to test it to support it)

Hi,

I wish to contribute to NTLM in reactive extension (To be specific in GitHub, refer quarkus-2.7\extensions\hibernate-reactive).

To test, you need a domain-based Windows login system (Laptop or Desktop or VM). This is not possible in my development environment because we do not have domain-based windows login in our company.
 
What we do/plan to do is 

- Deploy the application in the client UAT environment which employs domain-based windows login.
- Add a domain user in the SQL Server with "windows-authentication" activated.
- Configure the "reactive-url" in the application properties to include "domain name" and "integratedSecurity=true"
- Login into the client UAT environment using the domain user created in step 2
- Start the application

The application should start without any issues. The application should cater to data requests without any issues.

**Comment by temaleva** [comment]
I do not think it makes any sense to implement NTLM. It is complex, works only on windows and deprecated by Microsoft itself in favour of kerberos.
I would take a look on kerberos instead.
I am going to provide some more information on it shortly. Maybe this issue could be closed after that and new one for kerberos support is opened instead.

**Comment by temaleva** [comment]
So, to implement  SSPI authentication (no matter NTLM or Kerberos) LOGIN7 record should be sent  with SPNEGO packet instead of standard login. This SPNEGO (it calls Simple and Protected GSS-API Negotiation, but support kerberos and NTML as well ) packet includes SSPI Token. 
In best case server will respond with successful login, or it could be SSPI response and one additional step will be needed.

It would be perfect if at least SPNEGO packet is implemented in mssql  client, which is quite trivial  and lib user needs to provide a SSPI (GSS-API) token.

From the mssql client api point of view it would require a slight change in MSSQLConnection with additional token property and integrated authentication/security flag.

**Comment by tsegismont** [comment]
@temaleva thanks for your feedback. If I understand correctly, the code changes would be minimal: some config plus some modifications in the login packet exchange? Do you know how this could be at least manually tested, if not automatable?

**Comment by temaleva** [comment]
@tsegismont Yes, just additional payload added to LOGIN. 
And apart of this client code to prepare SSPI Token properly. 

Well, to be able to perform tests, Linux Container running SQL Server must be joined to AD Domain. (Microsoft has a plenty published docs on this point. e.g. this one:  https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-containers-ad-auth-adutil-tutorial?view=sql-server-ver17)
So for test there is a need to start Domain Controller VM (so real VM, docker container would be also possible, but not so easy) and SQL Server inside a container joined to Domain.
Domain controller should be possible with this VM appliance https://www.turnkeylinux.org/domain-controller.
